device/core/src/client_diagnostic.ts

```diff
@@ -0,0 +1,205 @@
+  private currentMessageNumber: number;
+
+  /**
+   * @constructor
+   */
+  constructor() {
+    // Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_001: [This constructor shall set sampling percentage to 0.]
+    this.diagSamplingPercentage = 0;
+    // Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_002: [This constructor shall set message number to 0.]
+    this.currentMessageNumber = 0;
+  }
+
+  /**
+   * @method            module:azure-iot-device.DiagnosticClient.getDiagSamplingPercentage
+   * @description       get the diagnostic sampling percentage.
+   *
+   * @returns {number}
+   */
+  getDiagSamplingPercentage(): number {
+    return this.diagSamplingPercentage;
+  }
+
+  /**
+   * @method            module:azure-iot-device.DiagnosticClient.setDiagSamplingPercentage
+   * @description       set the diagnostic sampling percentage.
+   *
+   * @param {number}   diagSamplingPercentage   The value of sampling percentage.
+   * @throws {ArgumentError}     If value < 0 or value > 100.
+   */
+  setDiagSamplingPercentage(diagSamplingPercentage: number): void {
+    if (typeof diagSamplingPercentage !== 'number') {
+      throw new errors.ArgumentError('Sampling percentage only accept number');
+    }
+    if (diagSamplingPercentage % 1 !== 0) {
+      throw new errors.ArgumentError('Sampling percentage should be integer');
+    }
+    // Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_003: [When percentage is less than 0 or larger than 100, throw IllegalArgumentException.]
+    if (diagSamplingPercentage < 0 || diagSamplingPercentage > 100) {
+      throw new errors.ArgumentError('Sampling percentage should be [0,100]');
+    }
+    this.diagSamplingPercentage = diagSamplingPercentage;
+    // Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_004: [This function shall reset message number to 0.]
+    this.currentMessageNumber = 0;
+  }
+
+  /**
+   * @method            module:azure-iot-device.DiagnosticClient.addDiagnosticInfoIfNecessary
+   * @description       add the diagnostic info
+   *
+   * @param {Message}   message   The message
+   */
+  addDiagnosticInfoIfNecessary(message: Message): void {
+    if (this.shouldAddDiagnosticInfo()) {
+      let diagnosticPropertyData: DiagnosticPropertyData = new DiagnosticPropertyData(this.generateEightRandomCharacters(), this.getCurrentTimeUtc());
+      message.diagnosticPropertyData = diagnosticPropertyData;
+    }
+  }
+
+  // Get a character from 0-9a-zA-Z
+  private getDiagnosticIdChar(value: number): string {
+    if (value <= 9) {
+      // 48 is char code of '0'
+      return String.fromCharCode(48 + value);
+    } else if (value <= 9 + 26) {
+      // 65 is char code of 'A'
+      return String.fromCharCode(65 + value - 10);
+    } else {
+      // 97 is char code of 'a'
+      return String.fromCharCode(97 + value - 36);
+    }
+  }
+
+  private generateEightRandomCharacters(): string {
+    // Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_005: [This function shall generate 8 random chars, each is from 0-9a-z.]
```

Comment:
>Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_005 [](start = 7, length = 39)

all these belong in a requirement document that we should review together prior implementation.